src/NServiceBus.Core/Licensing/LicenseManager.cs

```diff
@@ -2,37 +2,81 @@ namespace NServiceBus
 {
     using System;
     using System.Diagnostics;
-#if NETSTANDARD
-    using System.Runtime.InteropServices;
-#endif
     using System.Text;
     using System.Threading;
     using System.Threading.Tasks;
     using Logging;
     using Particular.Licensing;
+#if NETSTANDARD
+    using System.Runtime.InteropServices;
+
+#endif
 
     class LicenseManager
     {
         internal bool HasLicenseExpired => result?.HasExpired ?? true;
 
+        public LicenseManager(Func<DateTime> utcDateTimeProvider)
+        {
+            this.utcDateTimeProvider = utcDateTimeProvider;
+        }
+
         internal void InitializeLicense(string licenseText, string licenseFilePath)
         {
             var licenseSources = LicenseSources.GetLicenseSources(licenseText, licenseFilePath);
 
             result = ActiveLicense.Find("NServiceBus", licenseSources);
 
             LogFindResults(result);
-
             if (result.HasExpired)
             {
-                if (result.License.IsTrialLicense)
+                LogExpiredLicenseError(result.License, Logger);
+            }
+            else
+            {
+                LogExpiringLicenseWarning(result.License, Logger);
+            }
+
+            if (result.HasExpired && result.License.IsTrialLicense)
+            {
+                OpenTrialExtensionPage();
+            }
+        }
+
+        internal void LogExpiredLicenseError(License activeLicense, ILog logger)
+        {
+            if (activeLicense.UpgradeProtectionExpiration.HasValue)
+            {
+                logger.Error("Upgrade protection expired. Please extend your upgrade protection so that we can continue to provide you with support and new versions of the Particular Service Platform.");
+            }
+            else if (activeLicense.IsTrialLicense)
+            {
+                logger.Error("Trial license expired. Please extend your trial or purchase a license to continue using the Particular Service Platform.");
+            }
+            else
+            {
+                logger.Error("Platform license expired. Please extend your license to continue using the Particular Service Platform.");
+            }
+        }
+
+        internal void LogExpiringLicenseWarning(License activeLicense, ILog logger)
+        {
+            if (activeLicense.UpgradeProtectionExpiration.HasValue)
+            {
+                if (activeLicense.UpgradeProtectionExpiration.Value.Subtract(ExpirationWarningThreshold) <= utcDateTimeProvider().Date)
+                {
+                    logger.Warn("Upgrade protection expiring soon. Please extend your upgrade protection so that we can continue to provide you with support and new versions of the Particular Service Platform.");
```

Comment:
Instead of "soon", should we show the date?

Reply:
I was thinking about that. One problem is that the license is actually only considered expired after an additional day on the expiration date. Till that point, this warning will appear and I was thinking it would be looking weird. Also the expiration check is checking UTC times, so this could be kinda unexpected as well e.g. if your license is supposed to expire on the 12.12.2012 but you're in a timezone which is only in 11.12.2012 yet.

Reply:
I see, does that means that `result.HasExpired` is only `true` when `DateTime.UtcNow > UpgradeProtectionExpiration.AddDays(1)`? If so, that's quite confusing. Is there anything that can tell us when it will be true ahead of that time?

Regarding time zones, could we show both UTC and local to be explicit?

Reply:
the only way would be by copy&pasting from the existing licensing code. We can read the expiry dates but as said, the `HasExpired` property has a slightly different behavior instead of looking strictly at the expiry date alone. I've raised an issue to the licensing code to allow passing a current time instead of hardcoding it to `DateTime.UtcNow`.

Reply:
> Regarding time zones, could we show both UTC and local to be explicit?

tbh I think we should probably consider local timezones when checking license expiration. If I get a license which expires on 12.12.2012, I wouldn't expect the license to fail already hours before that date. But that's a different discussion imo.

Reply:
I'm not sure that's possible. E.g. a license holder may have servers spread across many time zones. Our licenses should probably expire at a specific UTC time, if they don't already, and that UTC time could be on different dates, depending on the observers local time zone.

Reply:
yeah, so there is definitely some details to consider. That's why I didn't really change anything from the current behavior. I was just trying to describe why I went with the currently proposed text.

Reply:
👍 makes sense.